mesonbuild/dependencies/base.py

```diff
@@ -476,16 +479,31 @@ def _set_libs(self):
             # file ourselves instead of depending on the compiler to find it
             # with -lfoo or foo.lib. However, we can only do this if we already
             # have some library paths gathered.
-            if self.static:
+            if try_static:
                 if lib.startswith('-L'):
                     libpaths.append(lib[2:])
                     continue
                 elif lib.startswith('-l') and libpaths:
                     args = self.compiler.find_library(lib[2:], self.env, libpaths, libtype='static')
-                    if not args or len(args) < 1:
+                    if args and not self.static:
+                        # A static lib exists but we are not forced to use it.
+                        # Check if it is found in an acceptable path, otherwise
+                        # keep -lfoo to use shared library.
+                        static_lib = args[0]
+                        for p in static_paths:
+                            if static_lib.startswith(p):
```

Comment:
Actually, revisiting a bit this patch, I think it can be greatly simplified. When static_paths is defined we could just not look at -L at all and replace libpaths by static_paths. That way self.compiler.find_library() will just find the .a files we want to keep.